Simbody/src/RigidBodyNodeSpec.cpp

```diff
@@ -552,6 +552,84 @@ RigidBodyNodeSpec<dof, noR_FM, noX_MB, noR_PF>::multiplyByMInvPass2Outward(
 }
 
 
+// Base to tip recursion for multiplyBySqrtMInv: 
+// temp allA_GB does not need to be initialized before
+// beginning the iteration.
+template<int dof, bool noR_FM, bool noX_MB, bool noR_PF> void
+RigidBodyNodeSpec<dof, noR_FM, noX_MB, noR_PF>::multiplyBySqrtMInvPass2Outward(
+    const SBInstanceCache&                  ic,
+    const SBTreePositionCache&              pc,
+    const SBArticulatedBodyInertiaCache&    abc,
+    const SBDynamicsCache&                  dc,
+    const Real*                             allEpsilon,
+    SpatialVec*                             allA_GB,
+    Real*                                   allUDot) const
+{
+    const Vec<dof>& eps  = fromU(allEpsilon);
+    SpatialVec&     A_GB = allA_GB[nodeNum];
+    Vec<dof>&       udot = toU(allUDot); // pull out this node's udot
+
+    const bool isPrescribed = isUDotKnown(ic);
+    const HType&        H   = getH(pc);
+    const PhiMatrix&    phi = getPhi(pc);
+    const Mat<dof,dof>& DI  = getDI(abc);
+
+    const HType&              G = getG(abc);
+    SpatialMat tauBar = G*~H;// 11*dof^2 flops 
+    SpatialMat psi = phi*tauBar; // ~100 flops 
+
+    int i, j, k;
+
+    Mat<dof,dof> sqrtDI(1);
+    // DI Cholesky Decomposition: 
+    if(dof>1){
+      double s1, s2;
+      Mat<dof,dof> L(0);
+      Mat<dof,dof> De(0);
+
+      for(j=0; j<dof; j++){
+        L(j,j) = 1.0;
+        // Diagonal
+        s1 = 0.0;
+        for(k=0; k<dof; k++){
+          s1 += (L(j,k) * L(j,k)) * De(k,k);
+        }
+        De(j,j) = DI(j,j) - s1;
+        // Off diagonal
+        for(i=(j+1); i<dof; i++){
+          s2 = 0.0;
+          for(k=0; k<j; k++){
+            s2 += L(i,k) * L(j,k) * De(k,k);
+          }
+          L(i,j) = (1 / De(j,j)) * (DI(i,j) - s2);
+        }
+      }
+
+      // Square root the diagonal
+      for(j=0; j<dof; j++){De(j,j) = sqrt(De(j,j));}
+
+      // Get LsqrtDe
+      sqrtDI = L *  De ;
+    }
+    else{
+      sqrtDI(0,0) = sqrt(DI(0,0));
+    }
+    // END Cholesky Decomposition
+
+    // Shift parent's acceleration outward (Ground==0). 12 flops
+    const SpatialVec& A_GP  = allA_GB[parent->getNodeNum()];
+    const SpatialVec  APlus = ~phi * A_GP;
```

Comment:
As you suggested and based on the changes made meanwhile in the code I renamed, the variables accordingly (f to v, A to V).
The algorithm is in Table 1 from Jain 2012 article  "Equipartition Principle for Internal Coordinate Molecular Dynamics". I put a detailed explanation with references in the function comments header in Simbody/include/simbody/internal/SimbodyMatterSubsystem.h
using the multiplyByMInv model. The code was derived from multiplyByMInv mostly apart from the Cholesky decomposition.
Regarding the units:
The input vector v can be thought of as a modal velocities vector with (kT)^1/2 =
([amu]^1/2 * [nm]) / [ps]) units and the resulting vector sqrtMinV will be a Maxwell-Boltzmann distribution scale parameter with velocity units that can be used to initialize velocities to a desired temperature.
Alternately, the input vector v can be thought of as unitless in order for the resulting vector A with [amu]^-1/2 units to be rescaled by a paramter with (kT)^1/2 units = ([amu]^1/2 * [nm]) / [ps]). After rescaling a vector with velocity units is obtained.